### PR TITLE
Implement GL(2,R)-orbit closure

### PR DESCRIPTION
**Pull request by saraedum**
and make it work over rings that are not fields such as ExactReals().

### CODE REVIEW

**Comment by saraedum** [comment]
@videlec @slel: This is a patch bomb sorry for that. It incorporates everything from #38 and changes a lot of code so things work with exact reals.

File: flatsurf/geometry/gl2r_orbit_closure.py [hunk 1]
```diff
@@ -0,0 +1,940 @@
...
+######################################################################
+
+import cppyy
+import gmpxxyy
+from pyeantic import RealEmbeddedNumberField
+
+import pyflatsurf
+import pyflatsurf.vector
+
+# TODO: move into flatsurf
```
Comment (saraedum) [inline flatsurf/geometry/gl2r_orbit_closure.py:44]: What is this one about @videlec

File: flatsurf/geometry/gl2r_orbit_closure.py [hunk 2]
```diff
@@ -0,0 +1,940 @@
...
+            sage: for d in O.decompositions(4, 20):  # optional: pyflatsurf
+            ....:     O.update_tangent_space_from_flow_decomposition(d)
+            ....:     if O.dimension() == 4:
+            ....:         break
+            sage: d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf
+            sage: dreal = d1/132 + d2/227 + d3/280 - d4/201  # optional: pyflatsurf
+            sage: dimag = d1/141 - d2/233 + d4/230 + d4/250  # optional: pyflatsurf
+            sage: d = [O.V2((x,y)).vector for x,y in zip(dreal,dimag)]  # optional: pyflatsurf
+
+        TODO: This is waiting for https://github.com/flatsurf/flatsurf/issues/145::
```
Comment (saraedum) [inline flatsurf/geometry/gl2r_orbit_closure.py:589]: This might work now…

**Comment by videlec** [comment]
The following example now generates a SEGFAULT
```
from flatsurf import polygons, translation_surfaces, similarity_surfaces
from flatsurf import GL2ROrbitClosure  # optional: pyflatsurf
S = translation_surfaces.mcmullen_genus2_prototype(4,2,1,1,0)
O = GL2ROrbitClosure(S)  # optional: pyflatsurf
u0,u1 = O.tangent_space_basis()  # optional: pyflatsurf
v0 = O.lift(u0)  # optional: pyflatsurf
v1 = O.lift(u1)  # optional: pyflatsurf
span([v0, v1])  # optional: pyflatsurf
```
[gdb traceback](https://github.com/flatsurf/sage-flatsurf/files/5169331/traceback.txt)

File: flatsurf/geometry/gl2r_orbit_closure.py (hunk 1 above)
Comment (videlec) [reply flatsurf/geometry/gl2r_orbit_closure.py:44]: I think that I wanted the following to work
```
from pylatsurf import Vertex
```
which is more convenient than
```
Vertex = cppyy.gbl.flatsurf.Vertex
```

File: flatsurf/geometry/gl2r_orbit_closure.py (hunk 2 above)
Comment (videlec) [reply flatsurf/geometry/gl2r_orbit_closure.py:603]: The example above this comment is now failing
```
from flatsurf import polygons, translation_surfaces, similarity_surfaces
from flatsurf import GL2ROrbitClosure
T = polygons.triangle(3,4,13)
S = similarity_surfaces.billiard(T)
S = S.minimal_cover("translation").erase_marked_points()
O = GL2ROrbitClosure(S)  # optional: pyflatsurf
for d in O.decompositions(4, 20):  # optional: pyflatsurf
    O.update_tangent_space_from_flow_decomposition(d)
    if O.dimension() == 4:
        break
d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf
```
gives
```
Traceback (most recent call last)
<ipython-input-2-603dc29281d6> in <module>()
      7     if O.dimension() == Integer(4):
      8         break
----> 9 d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf

<ipython-input-2-603dc29281d6> in <listcomp>(.0)
      7     if O.dimension() == Integer(4):
      8         break
----> 9 d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf

/home/vincent/programming/flatsurf/sage-flatsurf/flatsurf/geometry/gl2r_orbit_closure.py in lift(self, v)
    621         u = vector(self.V2.base_ring(), n + 1)
    622         u[:k] = v
--> 623         return A.solve_right(u)
    624 
    625     def absolute_homology(self):

/opt/sage/local/lib/python3.7/site-packages/sage/matrix/matrix2.pyx in sage.matrix.matrix2.Matrix.solve_right (build/cythonized/sage/matrix/matrix2.c:8345)()
    831 
    832         if not self.is_square():
--> 833             X = self._solve_right_general(C, check=check)
    834         else:
    835             try:

/opt/sage/local/lib/python3.7/site-packages/sage/matrix/matrix2.pyx in sage.matrix.matrix2.Matrix._solve_right_general (build/cythonized/sage/matrix/matrix2.c:9685)()
    954             # Have to check that we actually solved the equation.
    955             if self*X != B:
--> 956                 raise ValueError("matrix equation has no solutions")
    957         return X
    958 

ValueError: matrix equation has no solutions
```

**Comment by videlec** [comment]
Travis reports a doctest error
```
sage -t --long flatsurf/geometry/gl2r_orbit_closure.py

**********************************************************************

File "flatsurf/geometry/gl2r_orbit_closure.py", line 455, in flatsurf.geometry.gl2r_orbit_closure.GL2ROrbitClosure

Failed example:

    O  # optional: exactreal, pyflatsurf

Expected:

    GL(2,R)-orbit closure of dimension at least 11 in H_4(3, 1^3) (ambient dimension 11)

Got:

    GL(2,R)-orbit closure of dimension at least 9 in H_4(3, 1^3) (ambient dimension 11)

**********************************************************************
```

**Comment by saraedum** [comment]
Another things that's still broken here:
* [ ] for the torus, I now find an orbit-closure of dimension 3.

**Comment by videlec** [comment]
> Another things that's still broken here:
> 
>     * [ ]  for the torus, I now find an orbit-closure of dimension 3.

The dimension of the orbit-closure for a torus depends on the number of vertices (or marked points). The formula is `complex dim = 1 + # verts`.

**Comment by videlec** [comment]
After recompiling flint and dependencies, I get errors
```
sage: from flatsurf import polygons, translation_surfaces, similarity_surfaces                                                                                                                                      
sage: from flatsurf import GL2ROrbitClosure                                                                                                                                                                         
sage: S = translation_surfaces.mcmullen_genus2_prototype(4,2,1,1,0)                                                                                                                                                 
sage: O = GL2ROrbitClosure(S)                                                                                                                                                                                       
free(): invalid pointer

Thread 1 "python3" received signal SIGABRT, Aborted.
0x00007ffff7918615 in raise () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff7918615 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff7901862 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff795a5e8 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007ffff796227a in malloc_printerr () from /usr/lib/libc.so.6
#4  0x00007ffff796364c in _int_free () from /usr/lib/libc.so.6
#5  0x00007fffedd0622c in _fmpz_poly_gcd_subresultant (len2=0, poly2=0x55555f712e20, len1=3, poly1=0x55555f3d1c20, res=0x55555ffe1118) at gcd_subresultant.c:129
#6  _fmpz_poly_gcd_subresultant (res=res@entry=0x55555ffe1118, poly1=poly1@entry=0x55555f3d1c20, len1=len1@entry=3, poly2=poly2@entry=0x55555f712e20, len2=len2@entry=0) at gcd_subresultant.c:34
#7  0x00007fffedd03226 in _fmpz_poly_gcd (res=res@entry=0x55555ffe1118, poly1=poly1@entry=0x55555f3d1c20, len1=len1@entry=3, poly2=poly2@entry=0x55555f712e20, len2=len2@entry=0) at gcd.c:39
#8  0x00007fffedd40ea4 in _fmpq_poly_xgcd (G=G@entry=0x55555ffe1118, denG=denG@entry=0x55555ffe1128, S=S@entry=0x55555ffe1100, denS=denS@entry=0x55555ffe1110, T=T@entry=0x7fffffffa6b0, 
    denT=denT@entry=0x7fffffffa6c8, A=0x55555f3d1c20, denA=0x55555f127ce8, lenA=3, B=0x55555ffe1da0, denB=0x55555ffe1db8, lenB=0) at xgcd.c:83
#9  0x00007ffdc319994f in _nf_elem_inv (a=0x7fffffffa6b0, b=0x55555ffe1da0, nf=0x55555f127ce0) at inv.c:43
#10 0x00007ffdc3196565 in _nf_elem_div (a=a@entry=0x7fffffffa720, b=b@entry=0x55555821f020, c=c@entry=0x55555ffe1da0, nf=nf@entry=0x55555f127ce0) at div.c:26
#11 0x00007ffdc3196668 in nf_elem_div (a=a@entry=0x55555821f020, b=b@entry=0x55555821f020, c=c@entry=0x55555ffe1da0, nf=nf@entry=0x55555f127ce0) at div.c:41
#12 0x00007ffdc31d117b in renf_elem_div (a=0x55555821f020, b=0x55555821f020, c=0x55555ffe1da0, nf=0x55555f127ce0) at ./e-antic/renf_elem.h:445
#13 0x00007ffdc31d6b0f in std::function<void (renf_elem_struct*, renf_elem_struct const*, renf_elem_struct const*, renf*)>::operator()(renf_elem_struct*, renf_elem_struct const*, renf_elem_struct const*, renf*) const (__args#3=0x55555f127ce0, __args#2=<optimized out>, __args#1=<optimized out>, __args#0=<optimized out>, this=0x7fffffffa8d0) at /usr/include/c++/10.2.0/bits/std_function.h:622
#14 eantic::(anonymous namespace)::binop(eantic::renf_elem_class &, const eantic::renf_elem_class &, const std::function<void(renf_elem_struct*, const renf_elem_struct*, const renf_elem_struct*, renf*)> &) (
    lhs=..., rhs=..., op=...) at renfxx/renf_elem_class.cpp:27
#15 0x00007ffdc31d7c23 in eantic::renf_elem_class::operator/= (this=<optimized out>, rhs=...) at renfxx/renf_elem_class.cpp:566
#16 0x00007ffdc009b037 in ?? ()
#17 0x00007fffffffa9f0 in ?? ()
#18 0x000000005ffe3320 in ?? ()
#19 0x00007ffdc902abc0 in CppyyLegacy::gThreadTsd () from /opt/sage/local/lib/python3.7/site-packages/cppyy_backend/lib/libCoreLegacy.so
#20 0x0000000000000000 in ?? ()
```

**Comment by saraedum** [comment]
I assume you are using the latest e-antic?

**Comment by videlec** [comment]
On e-antic master at 701ce42

**Comment by saraedum** [comment]
I am actually on f95c10c3ab9d4250090398cc7b1411adabe2ed1d. Let me upgrade.

**Comment by saraedum** [comment]
I upgraded to the latest e-antic and the latest everything else in the flatsurf stack. I cannot reproduce that error.

File: .travis.yml [hunk 3]
```diff
@@ -19,11 +18,12 @@ before_install:
...
+  - mamba create -q -n test-environment -c flatsurf sagelib=$SAGE_VERSION pip pytest pytest-xdist surface-dynamics ipywidgets scipy matplotlib-base gap-defaults $CONDA_PACKAGES
   - conda activate test-environment
   - pip install $PIP_PACKAGES
   - pip install --verbose --no-index .
   # Show information message about cppyy regenerating pre-compiled headers so it does not show during the tests
   - python -c 'import cppyy' || true
 script:
-  - sage -t --force-lib --long --optional=$SAGE_OPTIONALS flatsurf doc
+  - sage -tp --force-lib --long --optional=$SAGE_OPTIONALS flatsurf doc
+  - pytest -n auto
```
Comment (videlec) [inline .travis.yml:29]: What is this `-n` option to pytest? It is not available in my conda setup...

Comment (videlec) [reply .travis.yml:29]: (the `--auto` is also not recognized for me)

**Comment by videlec** [comment]
pytest on my computer (conda setting) took very long time to run
```
42 passed, 1 warning in 2134.64s (0:35:34)
```
Looks much longer than what happend on travis.

File: flatsurf/geometry/gl2r_orbit_closure.py (hunk 1 above)
Comment (videlec) [reply flatsurf/geometry/gl2r_orbit_closure.py]: precised in 8d49414

File: flatsurf/geometry/gl2r_orbit_closure.py (hunk 2 above)
Comment (videlec) [reply flatsurf/geometry/gl2r_orbit_closure.py:609]: The problem to have this example works is that the `+` operation works in `pyflatsurf` surfaces and that `GL2ROrbitClosure` does not accept them. Is that something we want to allow? In order to allow that, one would need
- to be able to check whether the input `surface` in `GL2ROrbitClosure.__init__` is a `FlatTriangulation` (which is a template in cppyy, not something accepted by `isinstance`)
- extract the base ring from the `surface` in order to build the various vector spaces necessary for the `GL2ROrbitClosure` class

Comment (videlec) [reply flatsurf/geometry/gl2r_orbit_closure.py:609]: The alternative is of course to be able to go back and forth between `pyflatsurf` and `sage-flatsurf` in a more transparent fashion.

Comment (videlec) [reply flatsurf/geometry/gl2r_orbit_closure.py]: See also 4014171

File: .travis.yml (hunk 3 above)
Comment (saraedum) [reply .travis.yml:29]: It's from pytest-xdist which runs tests in parallel.

Comment (saraedum) [reply .travis.yml:29]: That probably also explains why tests are taking much longer for you than on travis.

**Comment by videlec** [comment]
There is something wrong with the `Decomposition.parabolic()` method in this branch or in the libflatsurf library
```
sage: T = polygons.triangle(1, 1, 3)
sage: S = similarity_surfaces.billiard(T)
sage: S = S.minimal_cover(cover_type="translation")
sage: O = GL2ROrbitClosure(S)
sage: for dec in O.decompositions_depth_first(5, 50):
sage:     assert dec.parabolic() == dec.decomposition.parabolic(), (dec.parabolic(), dec.decomposition.parabolic().value)
AssertionError: (False, 1)
```
When we erase the marked points the method `Decomposition.parabolic()` agrees with the parabolic from libflatsurf
```
sage: S = S.erase_marked_points()
sage: O = GL2ROrbitClosure(S)
sage: for dec in O.decompositions_depth_first(5, 50):
....:     assert dec.parabolic() == dec.decomposition.parabolic(), (dec.parabolic(), dec.decomposition.parabolic().value)
```

**Comment by videlec** [comment]
Some more stuff before merging
- [ ] `Decomposition` would better be the cppyy object with some plugged functions from Python. Methods that need the reference to the direction or the surface (`_spanning_tree_decomposition`, `kontsevich_zorich_cocycle`, `circumference_width` and `cylinder_deformation_subspace`) could be moved to `GL2ROrbitClosure`,
- [ ] `field_of_definition` is broken with exact reals (see #60).
And possibly think about
- [ ] `GL2ROrbitClosure` does not accept `pyflatsurf` surfaces as input (most probably in a further PR, see #59),
- [ ] display/conversion of boost tribool are annoying in Python... what could we do?

**Pull request merged**